src/algorithms/crypto/randomness.go

```diff
@@ -0,0 +1,40 @@
+package crypto
+
+import (
+	util "github.com/filecoin-project/specs/util"
+)
+
+type Randomness = util.Randomness
+type Serialization = util.Serialization
+
+// Derive a random byte string from a domain separation tag and an arbitrary
+// serializable object.
+//
+// Note: to produce values of type Serialization, use the auto-generated method prototypes
+//   Serialize_T(T) Serialization
+// for each type T defined in .id.
+//
+// In order to derive randomness from a collection of objects, rather than just a single
+// object, define a struct at the .id level that contains those objects as member fields.
+// This will then cause a Serialize_*() method to be generated for the struct type.
+func (tag DomainSeparationTag) DeriveRand(s Serialization) Randomness {
+	return _deriveRandInternal(tag, s, 0)
+}
+
+// As in DeriveRand(), but additionally accepts an index into the implicit pseudorandom stream.
+// Index must be strictly positive.
+func (tag DomainSeparationTag) DeriveRandExtended(s Serialization, index int) Randomness {
+	if index <= 0 {
+		panic("DeriveRandExtended only accepts indices > 0")
+	}
+	return _deriveRandInternal(tag, s, index)
+}
+
+func _deriveRandInternal(tag DomainSeparationTag, s Serialization, index int) Randomness {
+	buffer := []byte{}
```

Comment:
can you add a field delimeter between each field ('0' byte)?

Reply:
I'm not sure what purpose this would serve. The entries are already unambiguously delimited, by virtue of the fact that the first two are both constant-length (8 bytes). (And interposing a zero byte wouldn't help with ambiguity if they weren't already unambiguous -- e.g., some of the entries might just happen to have zero bytes in them anyway.)

Reply:
cc @whyrusleeping who likely won't review just yet. But you could try and find meaningful collisions by picking keys right/entries right. 

You are right they are constant for now but as a general rule... Costs nothing

Reply:
This doesn't quite make sense to me either. Backtracking a moment, just in case I misunderstood:

> can you add a field delimeter between each field ('0' byte)?

Does this mean adding two field delimiters to separate the three entries listed (i.e., tag, index, and serialization `s`), or does it mean adding field delimiters _within_ the serialization `s` itself?

Reply:
depends on how you build the serialization but i would say in both cases.

Reply:
I think it's important to separate the two cases out. In general, the property we'd like to prove is _unambiguity_. More precisely: for two inputs `x_0 = (tag_0, index_0, obj_0)` and `x_1 = (tag_1, index_1, obj_1)`, suppose `DeriveRand(x_i) = H(f(x_i))`. Then we want to prove:

> [Property A]:  If `x_0 != x_1`, then `f(x_0) != f(x_1)`.

from which the desired pseudorandomness will follow assuming we model the hash function H as a random oracle.

As defined above, we have

`f(tag, index, obj) = (tag || index || S(obj))`

where `S` denotes serialization of an arbitrary object. Since tag and index are both fixed-length, I claim property A holds for `f` trivially, _assuming_ property A holds for `S`. Moreover, if a serialization method `S` is functionally correct (meaning that it has a corresponding deserialization method that is its left-inverse), then this condition holds for `S`. Thus, we may take any serialization method `S` that is functionally correct under this definition.

In order to minimize redundancy, I suggest using the same serialization method `S` that we use elsewhere, i.e., currently CBOR-based serialization defined on algebraic datatypes. But I've left the specification parametric over `S`, in the sense that the `Serialize_T(T)` method prototypes auto-generated from .id files are still stubs at the moment. I think we should plan to update this in a future PR.

Reply:
Makes sound sense to me. Thanks for walking through that.